src/messages.ts

```diff
@@ -14,10 +14,48 @@
  * limitations under the License.
  */
 
-export const PROBLEMATIC_CREDENTIALS_WARNING =
-    `Your application has authenticated using end user credentials from Google
-  Cloud SDK. We recommend that most server applications use service accounts
-  instead. If your application continues to use end user credentials from Cloud
-  SDK, you might receive a "quota exceeded" or "API not enabled" error. For
-  more information about service accounts, see
-  https://cloud.google.com/docs/authentication/.`;
+import * as semver from 'semver';
+
+export function warn(warning: Warning) {
+  if (semver.satisfies(process.version, '>=8')) {
+    // @types/node doesn't recognize the emitWarning syntax which
+    // accepts a config object, so `as any` it is
+    // https://nodejs.org/docs/latest-v8.x/api/process.html#process_process_emitwarning_warning_options
+    // tslint:disable-next-line no-any
+    process.emitWarning(warning.message, warning as any);
+  } else {
+    // This path can be removed once we drop support for Node 6.
+    // https://nodejs.org/docs/latest-v6.x/api/process.html#process_process_emitwarning_warning_name_ctor
+    process.emitWarning(warning.message, warning.type);
+  }
+}
+
+export interface Warning {
+  code: string;
+  type: string;
+  message: string;
+}
+
+export const PROBLEMATIC_CREDENTIALS_WARNING = {
+  code: 'google-auth-library:DEP001',
+  type: 'DeprecationWarning',
```

Comment:
This is not a deprecation, just a plain warning. We are not removing or changing functionality. We are just letting users know that they might be using problematic default credentials.

With this in mind, maybe we should drop the `DEP` infix in the code? `google-auth-library:0001`?